tests/io/dask/dataframe/test_repartition.py

```diff
@@ -0,0 +1,131 @@
+import pandas as pd
+import pytest
+
+from kartothek.core.factory import DatasetFactory  # noqa: E402
+from kartothek.io.dask.dataframe import read_dataset_as_ddf
+from kartothek.io.dask.repartition import (
+    calculate_repartition_ratio,
+    optimize,
+    repartition_ddf,
+)
+from kartothek.io.iter import store_dataframes_as_dataset__iter
+from kartothek.io_components.metapartition import SINGLE_TABLE
+
+
+@pytest.fixture
+def dataset_factory_function_scope(store_factory):
+    return DatasetFactory(dataset_uuid="test", store_factory=store_factory)
+
+
+@pytest.mark.parametrize("frac", (1.0, 0.5, 0.1))
+@pytest.mark.parametrize(
+    "df,target_nrows_per_partition",
+    (  # repartition by decreasing partition size, increasing number of partitions
+        (pd.DataFrame({"A": range(100), "part": ["A", "B", "C", "D"] * 25}), 100),
+        (pd.DataFrame({"A": range(2 ** 7), "part": ["A", "B", "C", "D"] * 2 ** 5}), 8),
+        # repartition by increasing partition size, decreasing number of partitions
+        (pd.DataFrame({"A": range(50), "part": range(50)}), 10),
+        (pd.DataFrame({"A": range(10), "part": range(10)}), 10),
+    ),
+)
+def test_calculate_repartition_ratio_exact_match(
+    dataset_factory_function_scope, df, target_nrows_per_partition, frac
+):
+    """
+    Because we carefully choose the number of original partitions and the value of
+    `target_nrows_per_partition`, we get exact matches for `target_nrows_per_partition`.
+    This is not expected in a real-world scenario where the conditions are very unlikely.
+    """
+    dfs = [df]
+    store_dataframes_as_dataset__iter(
+        dfs,
+        store=dataset_factory_function_scope.store_factory,
+        dataset_uuid=dataset_factory_function_scope.dataset_uuid,
+        partition_on=["part"],
+    )
+
+    repartition_ratio = calculate_repartition_ratio(
+        dataset_factory=dataset_factory_function_scope,
+        table_name=SINGLE_TABLE,
+        dataset_metadata_frac=frac,
+        target_nrows_per_partition=target_nrows_per_partition,
+    ).compute()
+
+    ddf = read_dataset_as_ddf(factory=dataset_factory_function_scope)
+    ddf = repartition_ddf(ddf, repartition_ratio=repartition_ratio)
+
+    assert (ddf.map_partitions(len).compute() == target_nrows_per_partition).all()
+
+
+@pytest.mark.parametrize("frac", (0.3, 0.1))
+@pytest.mark.parametrize(
+    "df,target_nrows_per_partition",
+    (
+        # repartition by decreasing partition size, increasing number of partitions
+        (pd.DataFrame({"A": range(100), "part": ["A", "B", "C", "D"] * 25}), 97),
+        (pd.DataFrame({"A": range(100), "part": ["A", "B", "C", "D"] * 25}), 31),
+        # repartition by increasing partition size, decreasing number of partitions
+        (pd.DataFrame({"A": range(2 ** 6), "part": list("0123456789abcdef") * 4}), 11),
+        (pd.DataFrame({"A": range(10), "part": range(10)}), 7),
+        (pd.DataFrame({"A": range(10), "part": range(10)}), 1),
```

Comment:
These test cases also feel a bit redundant. Can we reduce this to a minimal set of cases we want to ensure work?

Reply:
Done. Please check and confirm.